diagnostic_tool/kinematics_cpp/kinematics.cpp

```diff
@@ -0,0 +1,443 @@
+/*
+ * Adapted from https://github.com/Le0nX/ModernRoboticsCpp
+ */
+#include <cmath>
+#include <vector>
+#include <Eigen/Dense>
+#include <pybind11/pybind11.h>
+#include <pybind11/eigen.h>
+
+namespace py = pybind11;
+
+#define M_PI           3.14159265358979323846  /* pi */
+
+namespace mr {
+
+/* Function: Find if the value is negligible enough to consider 0
+ * Inputs: value to be checked as a double
+ * Returns: Boolean of true-ignore or false-can't ignore
+ */
+    bool NearZero(const double val) {
+        return (std::abs(val) < .000001);
+    }
+
+/* Function: Returns the skew symmetric matrix representation of an angular velocity vector
+ * Input: Eigen::Vector3d 3x1 angular velocity vector
+ * Returns: Eigen::MatrixXd 3x3 skew symmetric matrix
+ */
+    Eigen::Matrix3d VecToso3(const Eigen::Vector3d &omg) {
+        Eigen::Matrix3d m_ret;
+        m_ret << 0, -omg(2), omg(1),
+                omg(2), 0, -omg(0),
+                -omg(1), omg(0), 0;
+        return m_ret;
+    }
+
+/*
+ * Function: Calculate the 6x6 matrix [adV] of the given 6-vector

```

Comment:
1. What is the 6-vector for here (e.g. an rotation + position vector).
2. It is probably well-known, but it is hard to tell what `ad` is for.

Reply:
Kept as is as much of the code is 3rd party source code.